File: training-data/lua/spiral_matrix.lua
local function spiral(n)
  local g = {}
  for r = 1, n do g[r] = {} end
  local top, left, bottom, right, k = 1, 1, n, n, 1
  while top <= bottom and left <= right do
    for c = left, right do g[top][c] = k; k = k + 1 end
    top = top + 1
    for r = top, bottom do g[r][right] = k; k = k + 1 end
    right = right - 1
    if top <= bottom then
      for c = right, left, -1 do g[bottom][c] = k; k = k + 1 end
      bottom = bottom - 1
    end
    if left <= right then
      for r = bottom, top, -1 do g[r][left] = k; k = k + 1 end
      left = left + 1
    end
  end
  return g
end

for _, row in ipairs(spiral(4)) do print(table.concat(row, "\t")) end

File: training-data/cpp/spiral_matrix.cpp
#include <iomanip>
#include <iostream>
#include <vector>

std::vector<std::vector<int>> spiral(int n) {
    std::vector<std::vector<int>> g(n, std::vector<int>(n));
    int top = 0, left = 0, bottom = n - 1, right = n - 1, k = 1;
    while (top <= bottom && left <= right) {
        for (int c = left; c <= right; ++c) g[top][c] = k++;
        ++top;
        for (int r = top; r <= bottom; ++r) g[r][right] = k++;
        --right;
        if (top <= bottom) { for (int c = right; c >= left; --c) g[bottom][c] = k++; --bottom; }
        if (left <= right) { for (int r = bottom; r >= top; --r) g[r][left] = k++; ++left; }
    }
    return g;
}

int main() {
    for (const auto& row : spiral(4)) {
        for (int v : row) std::cout << std::setw(3) << v;
        std::cout << '\n';
    }
}
